zingolib/src/wallet/tx_map_and_maybe_trees/trait_walletread.rs

```diff
@@ -0,0 +1,384 @@
+
+use secrecy::SecretVec;
+use shardtree::store::ShardStore;
+use zcash_client_backend::{
+    data_api::{Account, WalletRead},
+    keys::UnifiedFullViewingKey,
+};
+use zcash_primitives::consensus::BlockHeight;
+use zip32::AccountId;
+
+use crate::{error::ZingoLibError, wallet::notes::query::OutputQuery};
+
+use super::TxMapAndMaybeTrees;
+
+/// This is a facade for using LRZ traits. In actuality, Zingo does not use multiple accounts in one wallet.
+pub struct ZingoAccount(AccountId, UnifiedFullViewingKey);
+
+impl Account<AccountId> for ZingoAccount {
+    fn id(&self) -> AccountId {
+        self.0
+    }
+
+    fn source(&self) -> zcash_client_backend::data_api::AccountSource {
+        unimplemented!()
+    }
+
+    fn ufvk(&self) -> Option<&UnifiedFullViewingKey> {
+        Some(&self.1)
+    }
+
+    fn uivk(&self) -> zcash_keys::keys::UnifiedIncomingViewingKey {
+        unimplemented!()
+    }
+}
+
+/// some of these functions, initially those required for calculate_transaction, will be implemented
+impl WalletRead for TxMapAndMaybeTrees {
+    type Error = ZingoLibError;
+    type AccountId = AccountId;
+    type Account = ZingoAccount;
+
+    /// Returns the account corresponding to a given [`UnifiedFullViewingKey`], if any.
+    /// IMPL: partially implemented. zingo uses the default account. when we expand account functionality, this will be updated
+    fn get_account_for_ufvk(
+        &self,
+        ufvk: &UnifiedFullViewingKey,
+    ) -> Result<Option<Self::Account>, Self::Error> {
+        // todo we could assert that the ufvk matches, or return error.
+        Ok(Some(ZingoAccount(AccountId::ZERO, ufvk.clone())))
+    }
+
+    /// Returns the default target height (for the block in which a new
+    /// transaction would be mined) and anchor height (to use for a new
+    /// transaction), given the range of block heights that the backend
+    /// knows about.
+    ///
+    /// This will return `Ok(None)` if no block data is present in the database.
+    /// IMPL: fully implemented. the target height is always the next block after the last block fetched from the server, and the anchor is a variable depth below.
+    /// IMPL: tested
+    fn get_target_and_anchor_heights(
+        &self,
+        min_confirmations: std::num::NonZeroU32,
+    ) -> Result<
+        Option<(
+            zcash_primitives::consensus::BlockHeight,
+            zcash_primitives::consensus::BlockHeight,
+        )>,
+        Self::Error,
+    > {
+        match self.witness_trees.as_ref() {
+            Some(trees) => {
+                let opt_max_downloaded_height =
+                    match trees.witness_tree_orchard.store().max_checkpoint_id() {
+                        Ok(height) => height,
+                        Err(e) => match e {}, // Infallible
+                    };
+
+                Ok(opt_max_downloaded_height
+                    .map(|max_downloaded_height| max_downloaded_height + 1)
```

Comment:
Why not handle the option immediately?   Why carryit  from the first statement?

Reply:
.map IS handling the option

Reply:
I know.   But..   why let it get that far?